contracts/strategies/morpho/MetaMorphoStrategy.sol

```diff
@@ -149,11 +203,194 @@ contract MetaMorphoStrategy is ERC4626Strategy {
+      _pos = MORPHO_BLUE.position(_marketId, address(_mmVault));
+      // calc how much of `add` will be added to this market
+      if (_aprData.add > 0) {
+        _aprData.add = _calcMarketAdd(_mmVault, _marketId, _aprData.supplyLen, _aprData.add);
+      }
+      // calc how much of `sub` will be added to this market
+      if (_aprData.sub > 0) {
+        _aprData.sub = _calcMarketSub(_mmVault, _marketId, _aprData.withdrawLen, _aprData.sub);
+      }
+      // get underlyings supplied by the vault in the target market
+      // totalSupplyShares : totalSupplyAssets = supplyShares : assetsSuppliedByVault
+      // => assetsSuppliedByVault = supplyShares * totalSupplyAssets / totalSupplyShares
+      _assetsSuppliedByVault = _pos.supplyShares * _market.totalSupplyAssets / _market.totalSupplyShares + _aprData.add - _aprData.sub;
+      // calculate new totalSupplyAssets with liquidity added/removed
+      _totalSupplyAssets = _market.totalSupplyAssets + _aprData.add - _aprData.sub;
+      // calculate vaultShare (% in EXP_SCALE) of the total market and simulate change of liquidity by using add and sub
+      _vaultShare = _assetsSuppliedByVault * EXP_SCALE / _totalSupplyAssets;
+      // calculate vault rewards apr
+      apr = _rewardsInUnderlyings * _vaultShare * 100 / _totalSupplyAssets;
+    }
+  }
+
+  /// @notice quote rewards per year to underlyings
+  /// @param _rewardData reward data
+  function _quoteRewards(RewardData memory _rewardData) internal view returns (uint256) {
+    uint256 _oneReward = 10 ** uint256(IERC20Detailed(_rewardData.rewardToken).decimals());
+    // quote 1 reward to underlyings then multiply by supplyRewardTokensPerYear
+    return IStaticQuoter(UNI_V3_STATIC_QUOTER).quoteExactInput(_rewardData.uniV3Path, _oneReward) * 
+      IRewardsEmissions(MORPHO_REWARDS_EMISSION).rewardsEmissions(
+        _rewardData.sender,
+        _rewardData.urd,
+        _rewardData.rewardToken, 
+        _rewardData.marketId
+      ).supplyRewardTokensPerYear / _oneReward;
+  }
+
+  /// @notice calculate how much of vault `_add` amount will be added to this market
+  /// @param _mmVault metamorpho vault
+  /// @param _targetMarketId target market id
+  /// @param _supplyQueueLen supply queue length
+  /// @param _add amount of liquidity to add
+  function _calcMarketAdd(
+    IMMVault _mmVault,
+    bytes32 _targetMarketId,
+    uint256 _supplyQueueLen,
+    uint256 _add
+  ) internal view returns (uint256) {
+    IMorpho.Market memory _market;
+    // loop throuh supplyQueue, starting from the first market, and see how much will
+    // be deposited in target market
+    for (uint256 i = 0; i < _supplyQueueLen; i++) {
+      bytes32 _currMarketId = _mmVault.supplyQueue(i);
+      _market = MORPHO_BLUE.market(_currMarketId);
+      // get max depositable amount for this market
+      (uint184 _marketCap,,) = _mmVault.config(_targetMarketId);
+      uint256 _maxDeposit = uint256(_marketCap) - _market.totalSupplyAssets;
```

Comment:
good point

Reply:
Good catch you are right, I misinterpreted market cap as a cap on the market itself rather than the max amount depositable by the vault. Will update it

Reply:
btw I'm going to use the `_assetsSuppliedByVault` that I already have in the `_getRewardApr` (without `add` and `sub`) instead of fetching `expectedSupplyAssets` to avoid some gas overhead as the balance without the accrued interest is good enough for us for this method

Reply:
I have updated this but used the 'proportion' method instead of `expectedSupplyAssets` to reduce the gas usage a bit as this estimation should be good enough for this method https://github.com/Idle-Labs/idle-tranches/pull/85/commits/b190d3225ae6cd290939f4bbe9467edc0cea577b#diff-b2b538ff3955f5fdfdd12941175e16909cf6344c74ac7170fcc268062ef4920bR338-R339